.github/workflows/macos.yml

```diff
@@ -0,0 +1,81 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+name: macOS Build
+
+on:
+  push:
+  pull_request:
+
+permissions:
+  contents: read
+
+concurrency:
+  group: ${{ github.workflow }}-${{ github.repository }}-${{ github.head_ref || github.sha }}
+  cancel-in-progress: true
+
+jobs:
+  macos-build:
+    name: "${{ matrix.os }}"
+    strategy:
+      fail-fast: false
+      matrix:
+        # macos-13 = x86_64 Mac
+        # macos-14 = arm64 Mac
+        os: [macos-13, macos-14]
+    runs-on: ${{ matrix.os }}
+    env:
+      CCACHE_DIR: '${{ github.workspace }}/.ccache'
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
+        with:
+          submodules: recursive
+      - name: Install Dependencies
+        run: |
+          brew install \
+            bison boost ccache double-conversion flex fmt gflags glog \
+            icu4c libevent libsodium lz4 lzo ninja openssl range-v3 simdjson \
+            snappy thrift xz xsimd zstd
+          
+          echo "NJOBS=`sysctl -n hw.ncpu`" >> $GITHUB_ENV
+          
+      - name: Cache ccache
+        uses: actions/cache@v4
+        with:
+          path: '${{ env.CCACHE_DIR }}'
+          key: ccache-macos-${{ matrix.os }}-${{ hashFiles('velox/*') }}
```

Comment:
We used to use the merge base date for hashing, wont hasing based on files in velox/* be more noisy ?

Reply:
> be more noisy

In what way? On main we will have the same hash if no changes to `velox/` happen (pretty unlikely but hey) so the cash will be restored. If there were changes and a direct match for the key wasn't available it will do a partial match for the most recent cache matching `ccache-macos-macos-14*` and create a new cache after the build is complete. Using the merge day would increase the number of direct cache hits intra-day but will also prevent any cache updates after the first cache is created, because caches are immutable. This will in turn increase compile time due to ccache misses the more commits are made in a day.

For PRs I am pretty sure we will have misses anyway as forked PRs don't have access to base repo caches. But not 100% on that (I will test to confirm).

Reply:
I just tested restoring a specific key from a main cache and it works: https://github.com/facebookincubator/velox/actions/runs/7832729544/job/21372116061?pr=8707#step:2:15

So PRs will also partial match on main caches and use those to get started.

Reply:
Ok good to know, so this means builds should be a little bit faster than what we have now.